lib/widgets/autocomplete.dart

```diff
@@ -140,50 +216,52 @@ class _ComposeAutocompleteState extends State<ComposeAutocomplete> with PerAccou
             Text(label),
           ])));
   }
+}
+
+class TopicAutocomplete extends AutocompleteField<TopicAutocompleteQuery, TopicAutocompleteResult, String> {
+  const TopicAutocomplete({
+    super.key,
+    required this.streamId,
+    required ComposeTopicController super.controller,
+    required super.focusNode,
+    required this.contentFocusNode,
+    required super.fieldViewBuilder,
+  });
+
+  final FocusNode contentFocusNode;
+
+  final int streamId;
 
   @override
-  Widget build(BuildContext context) {
-    return RawAutocomplete<MentionAutocompleteResult>(
-      textEditingController: widget.controller,
-      focusNode: widget.focusNode,
-      optionsBuilder: (_) => _resultsToDisplay,
-      optionsViewOpenDirection: OptionsViewOpenDirection.up,
-      // RawAutocomplete passes these when it calls optionsViewBuilder:
-      //   AutocompleteOnSelected<T> onSelected,
-      //   Iterable<T> options,
-      //
-      // We ignore them:
-      // - `onSelected` would cause some behavior we don't want,
-      //   such as moving the cursor to the end of the compose-input text.
-      // - `options` would be needed if we were delegating to RawAutocomplete
-      //   the work of creating the list of options. We're not; the
-      //   `optionsBuilder` we pass is just a function that returns
-      //   _resultsToDisplay, which is computed with lots of help from
-      //   MentionAutocompleteView.
-      optionsViewBuilder: (context, _, __) {
-        return Align(
-          alignment: Alignment.bottomLeft,
-          child: Material(
-            elevation: 4.0,
-            child: ConstrainedBox(
-              constraints: const BoxConstraints(maxHeight: 300), // TODO not hard-coded
-              child: ListView.builder(
-                padding: EdgeInsets.zero,
-                shrinkWrap: true,
-                itemCount: _resultsToDisplay.length,
-                itemBuilder: _buildItem))));
+  ComposeTopicController get controller => super.controller as ComposeTopicController;
+
+  @override
+  AutocompleteIntent<TopicAutocompleteQuery>? autocompleteIntent() => controller.autocompleteIntent();
+
+  @override
+  TopicAutocompleteView initViewModel(BuildContext context) {
+    final store = PerAccountStoreWidget.of(context);
+    return TopicAutocompleteView.init(store: store, streamId: streamId);
+  }
+
+  void _onTapOption(BuildContext context, TopicAutocompleteResult option) {
+    final intent = autocompleteIntent();
+    if (intent == null) return;
+    final replacementString = option.topic;
+    controller.value = intent.textEditingValue.replaced(TextRange(
+      start: intent.syntaxStart,
+      end: intent.textEditingValue.text.length), replacementString);
```

Comment:
nit: `replacementString` gets kind of visually lost here, off after the end of a multi-line other expression

The formatting used in the similar existing method seems fine:
```suggestion
    controller.value = intent.textEditingValue.replaced(
      TextRange(
        start: intent.syntaxStart,
        end: intent.textEditingValue.text.length),
      replacementString,
    )
```